modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -3775,57 +3775,42 @@ private synchronized void scheduleNext() {
         }
 
         /** Stopping handler. */
-        public synchronized void stop() {
-            stopping = true;
-
-            if (active != null)
-                active.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
+        public void stop() {
+            synchronized (this) {
+                stopping = true;
+            }
 
             RemoteSnapshotFilesRecevier r;
 
             while ((r = queue.poll()) != null)
                 r.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
 
-            Set<RemoteSnapshotFilesRecevier> futs = activeTasks();
-            GridCompoundFuture<Void, Void> stopFut = new GridCompoundFuture<>();
-
-            try {
-                for (IgniteInternalFuture<Void> fut : futs)
-                    stopFut.add(fut);
-
-                stopFut.markInitialized().get();
-            }
-            catch (IgniteCheckedException e) {
-                throw new IgniteException(e);
-            }
+            if (active != null)
+                active.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
         }
 
-        /**
-         * @param nodeId A node left the cluster.
-         */
+        /** @param nodeId A node left the cluster. */
         public void onNodeLeft(UUID nodeId) {
-            Set<RemoteSnapshotFilesRecevier> futs = activeTasks();
             ClusterTopologyCheckedException ex = new ClusterTopologyCheckedException("The node from which a snapshot has been " +
                 "requested left the grid");
 
-            futs.forEach(t -> {
-                if (t.rmtNodeId.equals(nodeId))
-                    t.acceptException(ex);
+            queue.forEach(r -> {
+                if (r.rmtNodeId.equals(nodeId))
+                    r.acceptException(ex);
             });
-        }
-
-        /**
-         * @return The set of currently scheduled tasks, some of them may be already completed.
-         */
-        private Set<RemoteSnapshotFilesRecevier> activeTasks() {
-            Set<RemoteSnapshotFilesRecevier> futs = new HashSet<>(queue);
 
             RemoteSnapshotFilesRecevier active0 = active;
 
-            if (active0 != null)
-                futs.add(active0);
-
-            return futs;
+            if (active0 != null && !active0.isDone() && active0.rmtNodeId.equals(nodeId)) {
+                // If the task was started, but the remote node did not start the transmission and did not send SnapshotFilesFailureMessage.
+                // Otherwise, the task will fail via TransmissionHandler#onException.
```

Comment:
I think the `assert` in the `onException` is incorrect. Why `assert task.rmtNodeId.equals(nodeId);` ? We have single `SequentialRemoteSnapshotManager` for any transmission. WDYT?

Reply:
No, this is correct because only one transmission is allowed per node. All methods of the TransmissionHandler - about only one node for current transmission by design.

Reply:
> No, this is correct because only one transmission is allowed per node. All methods of the TransmissionHandler - about only one node for current transmission by design.

I cleary see that the assert still fires sometimes by the test. Everithing works, but the are a wierd stacktrace in this case.

Reply:
Added a check that `task.isDone()`. The tasks can be cancelled by stopChecker.